File: flipt-client-csharp/src/FliptClient/Models/ClientOptions.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace FliptClient.Models
{
    public class ClientOptions
    {
        /// <summary>
        /// Gets or sets the Flipt environment to use for evaluations. Defaults to 'default'.
        /// </summary>
        [JsonPropertyName("environment")]
        public string Environment { get; set; } = "default";

        /// <summary>
        /// Gets or sets the Flipt namespace to use for evaluations. Defaults to 'default'.
        /// </summary>
        [JsonPropertyName("namespace")]
        public string Namespace { get; set; } = "default";

        /// <summary>
        /// Gets or sets the Flipt API URL.
        /// </summary>
        [JsonPropertyName("url")]
        public string Url { get; set; } = "http://localhost:8080";

        /// <summary>
        /// Gets or sets request timeout as a duration (serialized as seconds).
        /// </summary>
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        [JsonPropertyName("request_timeout")]
        [JsonConverter(typeof(TimeSpanSecondsConverter))]
        public TimeSpan? RequestTimeout { get; set; }

        /// <summary>
        /// Gets or sets update interval as a duration (serialized as seconds).
        /// </summary>
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        [JsonPropertyName("update_interval")]
        [JsonConverter(typeof(TimeSpanSecondsConverter))]
        public TimeSpan? UpdateInterval { get; set; }

        /// <summary>
        /// Gets or sets authentication options.
        /// </summary>
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        [JsonPropertyName("authentication")]
        public Authentication? Authentication { get; set; }

        /// <summary>
        /// Gets or sets reference string for advanced use cases.
        /// </summary>
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        [JsonPropertyName("reference")]
        public string? Reference { get; set; }

        /// <summary>
        /// Gets or sets fetch mode for flag data (polling or streaming).
        /// </summary>
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        [JsonPropertyName("fetch_mode")]
        public FetchMode FetchMode { get; set; } = FetchMode.Polling;

        /// <summary>
        /// Gets or sets error strategy for evaluation failures.
        /// </summary>
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        [JsonPropertyName("error_strategy")]
        public ErrorStrategy ErrorStrategy { get; set; } = ErrorStrategy.Fail;

        /// <summary>
        /// Gets or sets the initial snapshot to use when instantiating the client.
        /// </summary>
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        [JsonPropertyName("snapshot")]
        public string? Snapshot { get; set; }

        /// <summary>
        /// Gets or sets TLS configuration for connecting to servers with custom certificates.
        /// </summary>
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        [JsonPropertyName("tls_config")]
        public TlsConfig? TlsConfig { get; set; }
    }

    [JsonConverter(typeof(LowercaseEnumConverter<ErrorStrategy>))]
    public enum ErrorStrategy
    {
        Fail,
        Fallback,
    }

    [JsonConverter(typeof(JsonStringEnumConverter))]
    public enum FetchMode
    {
        [JsonPropertyName("polling")]
        Polling,
        [JsonPropertyName("streaming")]
        Streaming,
    }

    public class Authentication
    {
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        [JsonPropertyName("client_token")]
        public string? ClientToken { get; set; }

        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        [JsonPropertyName("jwt_token")]
        public string? JwtToken { get; set; }
    }

    public class LowercaseEnumConverter<T> : JsonConverter<T>
        where T : struct, Enum
    {
        public override T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            var enumString = reader.GetString();
            if (Enum.TryParse<T>(enumString, true, out var value))
            {
                return value;
            }

            throw new JsonException($"Unable to convert \"{enumString}\" to {typeof(T)}.");
        }

        public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
        {
            writer.WriteStringValue(value.ToString().ToLower()); // Convert enum name to lowercase
        }
    }

    public class TimeSpanSecondsConverter : JsonConverter<TimeSpan?>
    {
        public override TimeSpan? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            if (reader.TokenType == JsonTokenType.Null)
            {
                return null;
            }

            if (reader.TokenType == JsonTokenType.Number && reader.TryGetInt32(out int seconds))
            {
                return TimeSpan.FromSeconds(seconds);
            }

            throw new JsonException($"Invalid value for TimeSpan: {reader.GetString()}");
        }

        public override void Write(Utf8JsonWriter writer, TimeSpan? value, JsonSerializerOptions options)
        {
            if (value.HasValue)
            {
                writer.WriteNumberValue((int)value.Value.TotalSeconds);
            }
            else
            {
                writer.WriteNullValue();
            }
        }
    }
}

File: flipt-client-kotlin-android/src/main/java/io/flipt/client/models/ClientOptions.kt
package io.flipt.client.models

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

/**
 * Serializes a [Duration] as a number of seconds (Long) for JSON.
 */
@kotlinx.serialization.ExperimentalSerializationApi
object DurationSecondsSerializer : KSerializer<Duration?> {
    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("DurationSeconds", PrimitiveKind.LONG)

    override fun serialize(
        encoder: Encoder,
        value: Duration?,
    ) {
        if (value == null) {
            encoder.encodeNull()
        } else {
            encoder.encodeLong(value.inWholeSeconds)
        }
    }

    override fun deserialize(decoder: Decoder): Duration? {
        return try {
            val seconds = decoder.decodeLong()
            seconds.seconds
        } catch (e: Exception) {
            null
        }
    }
}

/**
 * Configuration options for FliptClient.
 * @property environment The environment for the Flipt server (optional).
 * @property namespace The namespace for the Flipt server (optional).
 * @property url The URL for the Flipt server (optional).
 * @property requestTimeout The request timeout (optional, as Duration).
 * @property updateInterval The update interval (optional, as Duration).
 * @property authentication The authentication strategy (optional).
 * @property reference The reference for the Flipt server (optional).
 * @property fetchMode The fetch mode for the Flipt server (optional).
 * @property errorStrategy The error strategy (optional).
 * @property snapshot The initial snapshot for evaluation (optional).
 */
@Serializable
data class ClientOptions(
    @SerialName("environment") val environment: String? = null,
    @SerialName("namespace") val namespace: String? = null,
    @SerialName("url") val url: String? = null,
    @SerialName("request_timeout") @Serializable(with = DurationSecondsSerializer::class) val requestTimeout: Duration? = null,
    @SerialName("update_interval") @Serializable(with = DurationSecondsSerializer::class) val updateInterval: Duration? = null,
    @SerialName("authentication") val authentication: AuthenticationStrategy? = null,
    @SerialName("reference") val reference: String? = null,
    @SerialName("fetch_mode") val fetchMode: FetchMode? = null,
    @SerialName("error_strategy") val errorStrategy: ErrorStrategy? = null,
    @SerialName("snapshot") val snapshot: String? = null,
)
